R/staticServer.R

```
@@ -0,0 +1,80 @@
+#' Serve a directory
+#'
+#' Run a server to host a static directory.
+#'
+#' @examplesIf interactive()
+#' website_dir <- system.file("example-static-site", package = "httpuv")
+#' runStaticServer(dir = website_dir)
+#'
+#' @param dir The directory to serve. Defaults to the current working directory.
+#' @inheritParams startServer
+#' @param background Whether to run the server in the background. By default,
+#'   the server runs in the foreground and blocks the R console. You can stop
+#'   the server by interrupting it with `Ctrl + C`.
+#'
+#'   When `background = TRUE`, the server will run in the background and will
+#'   process requests when the R console is idle. To stop a background server,
+#'   call [stopAllServers()] or call [stopServer()] on the server object
+#'   returned (invisibly) by this function.
+#'
+#' @seealso [runServer()] provides a similar interface for running a dynamic
+#'   app server. Both `runStaticServer()` and [runServer()] are built on top of
+#'   [startServer()], [service()] and [stopServer()]. Learn more about httpuv
+#'   servers in [startServer()].
+#'
+#' @export
+runStaticServer <- function(
+  dir = getwd(),
+  host = "127.0.0.1",
+  port = 7446,
+  background = FALSE,
+  browse = interactive()
+) {
+  stopifnot(
+    "`port` must be an integer" = is.numeric(port),
+    "`port` must be an integer" = port == as.integer(port),
+    "`port` must be a single integer" = length(port) == 1
+  )
+
+  browse <- isTRUE(browse)
+  port <- requested_or_random_port(port, host)
```

I don't love it that a user can explicitly ask for a certain port and get a different one back. I'd prefer either just failing if the port is in use (even if it's the default port), or at most, only falling back to a random port if `missing(port)`.

---

Agreed, this bugged me too. I'll use a different pattern where we'll use a random port only if the default port is unavaiable and the user didn't ask for a specific port.